### ISSUE TITLE
possibility of data not being sent - expected behavior?

### ISSUE THREAD

**Issue by bohnman** [awslabs/amazon-kinesis-agent#36]
I have a server where we are using the agent to send log files to a kinesis streams.  Last night, I had to reboot the server.  What I didn't realize was that the agent wasn't set to start on boot until about 20 minutes later.  

After I started up the agent, my expectation was that agent would start from the last file offset it had recorded in the SQL Lite database.  However, it seems that agent started at the end of the file, which resulted in a 20 minute gap of the data.

I see that there's initialPosition, but I wouldn't want to set it to START_OF_FILE because I wouldn't want to re-process a large number of event needlessly.

Is there a way for the agent to continue where it left off after restart or do you have to roll your own solution?

**Comment by chaochenq** [comment]
Hi,

The expected behavior is to catch up from the last checkpoint:

https://github.com/awslabs/amazon-kinesis-agent/blob/master/src/com/amazon/kinesis/streaming/agent/tailing/FileTailer.java#L142

Could you look at the agent log to see if there's any indication upon starting the agent? e.g.

"Successfully started tailing from previous checkpoint"
or
"Failed to start tailing from previous checkpoint"

**Comment by bohnman** [comment]
I didn't see anything in the logs.  I have attached both the agent.json config and some of the agent log output.

The way we log to files is to write out to a file with the UTC date in it so we get automatically rolling.  Here's a directory listing of our application log directory:

```
[rbohn@cloud-ingest-prod ingest-server]$ ls -ltr
total 2448220
-rw-rw-r-- 1 bht bht 350156792 Jun 25 19:59 events.2016-06-25-utc.json
-rw-rw-r-- 1 bht bht 345300096 Jun 26 19:59 events.2016-06-26-utc.json
-rw-rw-r-- 1 bht bht 345737121 Jun 27 19:59 events.2016-06-27-utc.json
-rw-rw-r-- 1 bht bht 342976405 Jun 28 19:59 events.2016-06-28-utc.json
-rw-rw-r-- 1 bht bht 366632717 Jun 29 19:59 events.2016-06-29-utc.json
-rw-rw-r-- 1 bht bht 392003010 Jun 30 19:59 events.2016-06-30-utc.json
drwxr-xr-x 2 bht bht      4096 Jun 30 23:13 staging
drwxr-xr-x 2 bht bht     12288 Jun 30 23:13 archive
-rw-rw-r-- 1 bht bht 364083687 Jul  1 18:38 events.2016-07-01-utc.json
```

[agent-log.txt](https://github.com/awslabs/amazon-kinesis-agent/files/344246/agent-log.txt)
[agent.json.txt](https://github.com/awslabs/amazon-kinesis-agent/files/344251/agent.json.txt)

**Comment by chaochenq** [comment]
Interesting... I just tried to reproduce this. So I manually killed the agent process, waited for 2 minutes, and started the agent again. I was able to see that my agent started tailing from the last checkpoint:


2016-07-01 17:21:25.317-0700 chaocheq.desktop.amazon.com (Agent.MetricsEmitter RUNNING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Progress: 1518
1 records parsed (3974360 bytes), and 14849 records sent successfully to destinations. Uptime: 330294ms
2016-07-01 17:21:25.317-0700 chaocheq.desktop.amazon.com (Agent.MetricsEmitter RUNNING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Tailing is 0.0
00977 MB (1507 bytes) behind.
2016-07-01 17:23:41.831-0700 chaocheq.desktop.amazon.com (main) com.amazon.kinesis.streaming.agent.Agent [INFO] Reading configuration from file: /rhel5pdi/wo
rkplace/chaocheq/cloud9-dev/AWSFirehoseAgent/src/AWSFirehoseAgent/support/aws-kinesis-agent.json
2016-07-01 17:23:41.868-0700 chaocheq.desktop.amazon.com (main) com.amazon.kinesis.streaming.agent.Agent [INFO] null: Agent will use up to 24 threads for sen
ding data.
2016-07-01 17:23:41.982-0700 chaocheq.desktop.amazon.com (main) com.amazon.kinesis.streaming.agent.tailing.checkpoints.SQLiteFileCheckpointStore [INFO] Delet
ed 0 old checkpoints.
2016-07-01 17:23:41.988-0700 chaocheq.desktop.amazon.com (Agent STARTING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Starting up...
2016-07-01 17:23:41.994-0700 chaocheq.desktop.amazon.com (Agent STARTING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Starting tailer for file fh:chaocheqfirehose-us-west-2:/tmp/aws-kinesis-agent-test1.log*
2016-07-01 17:23:42.040-0700 chaocheq.desktop.amazon.com (FileTailer[fh:chaocheqfirehose-us-west-2:/tmp/aws-kinesis-agent-test1.log*]) com.amazon.kinesis.streaming.agent.tailing.FirehoseParser [INFO] FirehoseParser[fh:chaocheqfirehose-us-west-2:/tmp/aws-kinesis-agent-test1.log*]: Opening /tmp/aws-kinesis-agent-test1.log for parsing.
2016-07-01 17:23:42.040-0700 chaocheq.desktop.amazon.com (FileTailer[fh:chaocheqfirehose-us-west-2:/tmp/aws-kinesis-agent-test1.log*]) com.amazon.kinesis.streaming.agent.tailing.FileTailer [INFO] FileTailer[fh:chaocheqfirehose-us-west-2:/tmp/aws-kinesis-agent-test1.log*]: Successfully started tailing from previous checkpoint /tmp/aws-kinesis-agent-test1.log@3269236
2016-07-01 17:23:42.041-0700 chaocheq.desktop.amazon.com (Agent STARTING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Starting tailer for file kinesis:kinesis-agent-test2:/tmp/aws-kinesis-agent-test2.log*
2016-07-01 17:23:42.046-0700 chaocheq.desktop.amazon.com (FileTailer[kinesis:kinesis-agent-test2:/tmp/aws-kinesis-agent-test2.log*]) com.amazon.kinesis.streaming.agent.tailing.KinesisParser [INFO] KinesisParser[kinesis:kinesis-agent-test2:/tmp/aws-kinesis-agent-test2.log*]: Opening /tmp/aws-kinesis-agent-test2.log for parsing.
2016-07-01 17:23:42.046-0700 chaocheq.desktop.amazon.com (FileTailer[kinesis:kinesis-agent-test2:/tmp/aws-kinesis-agent-test2.log*]) com.amazon.kinesis.streaming.agent.tailing.FileTailer [INFO] FileTailer[kinesis:kinesis-agent-test2:/tmp/aws-kinesis-agent-test2.log*]: Successfully started tailing from previous checkpoint /tmp/aws-kinesis-agent-test2.log@618414
2016-07-01 17:23:42.051-0700 chaocheq.desktop.amazon.com (Agent STARTING) com.amazon.kinesis.streaming.agent.Agent [INFO] Agent: Startup completed in 62 ms.

**Comment by chaochenq** [comment]
I `sudo kill -9` the process at 17:21, so you won't see the ShutDown message in the log. but you can see it reading upon the last checkpoint after starting up at 17:23

**Comment by chaochenq** [comment]
In your case, possibly it didn't find the checkpoint for kinesis:event-router:/opt/bht/data/ingest-server/events*.json? Could you check if there's entry in the checkpoint file under /var/run/aws-kinesis-agent/

**Comment by chaochenq** [comment]
Yes. Please add "log.level" : "DEBUG" to your agent config file and look for the debug messages during the initialization of the tailer:

https://github.com/awslabs/amazon-kinesis-agent/blob/master/src/com/amazon/kinesis/streaming/agent/tailing/FileTailer.java#L136

I wonder if it was due to agent failed to find or load the checkpoint for some reason, and went ahead to tail from the end of file

**Comment by chaochenq** [comment]
Can this PR possibly fix the issue?

https://github.com/awslabs/amazon-kinesis-agent/pull/38

**Comment by bohnman** [comment]
I doubt it.  /var/run is not on tmpfs so permissions should have been preserved.

Filesystem      Size  Used Avail Use% Mounted on
/dev/xvda1      128G   13G  116G  10% /
devtmpfs        994M   56K  994M   1% /dev
tmpfs          1002M   12K 1002M   1% /dev/shm

**Comment by bohnman** [comment]
I figured it out.  I'm running Amazon Linux on EC2 and at least the version I'm running, clears out files in /var/run on reboot.

I had to add the following line to the agent.json:

`"checkpointFile": "/var/aws-kinesis-agent/checkpoints",`

After I did this, I rebooted and the agent started resuming from the previous checkpoint.

Given that a lot of linux distros clear out /var/run, ou may want to consider changing the default location of the checkpoints file
